features/FxA-83-sign-in-confirmation-email/README.md

```diff
@@ -0,0 +1,290 @@
+armed with a list of known email/password pairs
+to maliciously access users’ Firefox Accounts.
+In such cases,
+attackers would have full access
+to the user’s Sync data,
+including passwords, history, bookmarks and
+the ability to force-install malicious add-ons.
+
+## Outcomes
+
+If an attack occurs,
+we would like fewer accounts to be compromised
+without negatively impacting
+the number of signed-in users
+or our overall FxA engagement rate.
+
+## Hypothesis
+
+If we introduce an additional email confirmation step
+for Sync sign-ins,
+fewer accounts will be compromised.
+
+We will know this to be true if,
+in the event of an attack,
+the number of successful sign-ins remains constant
+and the count of 104 (unverified account) errors
+from protected endpoints
+increases in line with any wider increase in traffic.
+Further, we can measure that
+the confirmation step is not affecting
+legitimate users
+by monitoring the success rate
+of connecting to Sync
+before and after deployment.
+
+## User stories
+
+* As a Sync user,
+  when signing in to my Firefox Account,
+  I want to confirm my identity
+  via my email account
+  and I want my account access to be limited
+  until my identity has been confirmed.
+
+* As a user of an OAuth relier,
+  when signing in to my Firefox Account,
+  I do not want an extra confirmation step.
+
+## Constraints
+
+* In light of the recent attack,
+  the solution should be quick to implement.
+
+* The solution must work across all clients,
+  without patches landing in client code.
+
+* The solution must not affect existing sessions
+  from the user's point of view.
+
+* The solution must work from
+  regular Sync sign-in,
+  `/force_auth` and
+  when signing-in to Sync from `/signup`.
+
+## Proposed solution
+
+Add a notion of verification
+to sessionTokens and keyFetchTokens.
+Tokens that are unverified
+have reduced powers (see below).
+Token verification is achieved
+by following a link sent by email.
+
+### What happens when a Sync user signs in?
+
+1. User submits form.
+
+2. Content server requests `POST /account/login?keys=true`.
+
+3. Auth server generates `tokenVerificationId`
+   and sends it with `tokenVerified:false`
+   in request to `PUT /sessionToken/:tokenId`.
+
+4. Auth server sends same `tokenVerificationId`
+   and `tokenVerified:false`
+   in request to `PUT /keyFetchToken/:tokenId`.
+
+5. Auth server invokes the mailer
+   to send a confirmation email,
+   which includes `tokenVerificationId` in the URL.
+
+6. Auth server sends response back to content server,
+   including in the data:
+   ```
+   {
+     "sessionToken": sessionTokenId,
+     "keyFetchToken": keyFetchTokenId,
+     "challengeReason": "signin",
+     "challengeMethod": "email"
+   }
+   ```
+
+7. Content server sends web channel message
+   notifying the browser of successful sign-in.
+
+8. Content server navigates to confirm-signin view.
+
+9. Content server starts polling
+   on `GET /token/:tokenId/status`.
```

Comment:
Ewww, does it? Consistency is important I guess, but `POST` seems like an odd choice from a semantic point of view.